woocommerce/sourcebuster-js
# Add more web storage options

This PR enables the user to specify which type of storage to use via the `web_storage` configuration parameter:
- `web_storage: 'cookies'` (default) - multiple cookies as now
    - Proven implementation. 
    - Some users don't like having so many cookies (extra requests, etc).
    - Some cookies trigger a false positive for Web Application Firewalls (like ModSecurity) using the Comodo WAF ruleset.
- `web_storage: 'singleCookie'` - one JSON-encoded `sbjs_current` cookie that contains all data stored in the cookies
    - Workaround for Web Application Firewall 403 errors which generally have `sbjs_current` whitelisted.
    - Includes a method for handling session length (able to count user sessions)
    - Reduces the number of cookies passed back and forth to the server.
    - Limited to 4kb in data, which could be a problem if there are many 4-byte values (emoji, Sumerian, Hieroglyphics, etc.)
- `web_storage: 'sessionStorage'` - keep all data in the browser's session storage
    - No cookies
    - All data expires once the session ends (no counting user sessions)
    - Data isn't synced across multiple browser windows (opening a link in a new tab loses the data)
- `web_storage: 'localStorage'` - keep all the data in the browser's local storage
    - No cookies
    - Data will persist indefinitely
    - Includes a method for handling session length (able to count user sessions)
    - Data will persist indefinitely

##### Notes
- This doesn't currently handle moving from one storage type to another. If a user visits the site with one storage method configured, then the configuration changes during their session or while navigating, the original data will be lost (at best) or corrupted.
- Please reread the sub-bullet points above as they illustrate pros and cons of each method!
- The `localStorage` uses the maximum of the session_length (minutes) and lifetime (months converted to minutes) to create an expiry date for the values. Technically they'll live "forever" in the browser's local storage, but whenever Sourcebuster tries to access them, if the `expiry` timestamp has passed, the values will be discarded. This workaround was added to prevent 0-lifetime configuration (all cookies expire with the session) from creating 0-lifetime local storage, which effectively is erased and set every time a page is loaded.

### Testing
1. Clone the repo into a web-accessible directory.
2. Navigate to `/public/index.html` in a browser.
3. At the bottom of the `index.html` file, find the Sourcebuster JavaScript configuration. Initially, it should have `web_storage` set as `cookies`.
4. You can add a `session_length` parameter with value of `1` minute if you would like to test transitions from session to session (default is `30` minutes).
5. Perform some testing with the `cookies` value for `web_storage`. 
    - Ensure that Referral (coming from [woo.com](https://woo.com)), Organic (coming from google.com), TypeIn/Direct, UTM (with `?utm_source=[VALUE]…`)  all capture source values correctly.
    - Change source method and see that the "first" source values and "current" source values are correct (UTM parameters will always override "first").
    - Check for incrementing pages viewed in the session.
    - Wait for session timeout then refresh, and see pages viewed return to `1` and user sessions increase.
    - Confirm multiple cookies in the Application tab of the dev panel.
6. Close the browser (or erase all `sbjs_` cookies) and change the `web_storage` config value to `singleCookie` (camel cased).
7. Run the above tests again and verify that the behavior is consistent with the previous tests.
8. Also, check in the Application tab in the dev panel to confirm that only one cookie is being created.
9. Repeat the tests with both `localStorage` and `sessionStorage` options.
    - Be sure to close the browser (and/or erase previous config values in storage or cookies) as the current implementation doesn't erase any old data when changing the `web_storage` value.
    - Confirm that no cookies are created when those these methods are enabled.
